src/utils.cc

```diff
@@ -140,7 +140,10 @@ std::string realPath(const std::string &path) {
 bool normalizeFolder(std::string &path) {
   for (auto &[root, real] : g_config->workspaceFolders)
     if (real.size() && llvm::StringRef(path).startswith(real)) {
-      path = root + path.substr(real.size());
+      // extra memory allocation may happen when string concatenates
```

Comment:
```suggestion
      path.reserve(root.size() + path.size() - real.size());
```
why not reserve upfront since we already know the size?